lib/Controller/WhiteboardController.php

```diff
@@ -106,4 +85,52 @@ public function show(int $fileId): DataResponse {
 			'data' => $data,
 		]);
 	}
+
+	private function verifySharedToken(string $token, int $fileId): bool {
+		[$roomId, $timestamp, $signature] = explode(':', $token);
+
+		if ($roomId !== (string)$fileId) {
+			return false;
+		}
+
+		$sharedSecret = $this->configService->getWhiteboardSharedSecret();
+		$payload = "{$roomId}:{$timestamp}";
+		$expectedSignature = hash_hmac('sha256', $payload, $sharedSecret);
+
+		return hash_equals($expectedSignature, $signature);
+	}
+
+	/**
+	 * @throws NotPermittedException
+	 * @throws NoUserException
+	 * @throws JsonException
+	 */
+	#[NoAdminRequired]
+	#[NoCSRFRequired]
+	#[PublicPage]
+	public function update(int $fileId, array $data): DataResponse {
+		$whiteboardAuth = $this->request->getHeader('X-Whiteboard-Auth');
+		$whiteboardUser = $this->request->getHeader('X-Whiteboard-User');
+
+		if (!$whiteboardAuth || !$this->verifySharedToken($whiteboardAuth, $fileId)) {
+			return new DataResponse(['message' => 'Unauthorized'], Http::STATUS_UNAUTHORIZED);
+		}
+
+		$user = $this->userManager->get($whiteboardUser);
+
+		if (!$user) {
+			return new DataResponse(['message' => 'Invalid user'], Http::STATUS_BAD_REQUEST);
+		}
+
```

Comment:
We probably want to call [`setVolatileActiveUser`](https://github.com/nextcloud/server/blob/dae7c159f728a90ffa53247d6e033abdae5d2bd6/lib/public/IUserSession.php#L51) to make sure the write is also in scope of the user (e.g. for creating versions and activity entries).